Add an endpoint to create a terminal intent

anu-rock opened the pull request:
Fixes #3431

#### Changes proposed in this Pull Request

A new endpoint that leverages an existing API method to create a server-side intent for an in-person (or terminal) payment. It is specific to terminal payments and DOES NOT offer a way to create an open-ended payment intent.

#### Testing instructions

* Create a new order in WooCommerce with "Cash on delivery" as the payment method.
* Make a POST request to `/payments/orders/<order_id>/create_terminal_intent`. Refer the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.
* You should get a successful (200) response with the id of the created intent.
* Try again with a non-existent order id. In this case, you should receive a 404 response with an appropriate error message.

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt`
- [x] Covered with tests
- [ ] Tested on mobile (does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (does not apply)

anu-rock committed:
Add an endpoint to create a terminal intent

anu-rock commented:
Questions:

* Is there any metadata that should be recorded while creating an intent?
* Any other data about the intent (apart from id) that should be returned in response?

dpaun1985 commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,49 @@ public function create_customer( $request ) {
+			// Create a new intention.
+			$result = $this->gateway->create_intent( $order, $payment_methods, $capture_method );
+			if ( 'created' !== $result['status'] ) {
+				$http_code = $result['http_code'] ?? 502;
```
it seems that we always send `http_code` . we can skip this check.

anu-rock replied:
Good point. Done ✅

dpaun1985 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,58 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+	 *
+	 * @throws Exception - When an error occurs in intent creation.
+	 */
+	public function create_intent( WC_Order $order, array $payment_methods, string $capture_method ) {
```
you could add the default value of `$capture_method` here and remove it from the method body.

anu-rock replied:
Nice. Done ✅

dpaun1985 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,58 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+	public function create_intent( WC_Order $order, array $payment_methods, string $capture_method ) {
+		$amount           = $order->get_total();
+		$currency         = strtolower( $order->get_currency() );
+		$converted_amount = WC_Payments_Utils::prepare_amount( $amount, $currency );
```
since the amount is used only here, you could use `$order->get_total()` and skip the `$amount` declaration

anu-rock replied:
The reason `$amount` exists is to annotate my code better without comments 😄 This way, it's instantly understandable by a new developer what the first argument of `prepare_amount` is. It also makes the code a bit cleaner.

anu-rock replied:
> Anyways, in case of prepare_amount with its only two arguments and one being self-explanatory, I think you can safely omit $account assignment.

Owing to the consensus, I've made the change 😄

anu-rock replied:
> I also find it outrageous that WC_Payments_Utils::prepare_amount doesn't do case conversion if it needs them and requires caller to take care of it. 👴

Indeed +1

dpaun1985 commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,49 @@ public function create_customer( $request ) {
+
+			// Create a new intention.
+			$result = $this->gateway->create_intent( $order, $payment_methods, $capture_method );
+			if ( 'created' !== $result['status'] ) {
```
in order to avoid adding new status variable with different values, we could catch `API_Exception` here.

anu-rock replied:
I'm not sure if I follow. Could you give an example?

dpaun1985 replied:
I mean you can let `create_intent` to throw `API_Exception` ( now you catch the exception inside the method) and catch it here . this way you can avoid introducing status 'created'/'failed'.

anu-rock replied:
I've refactored it as per this comment:  
https://github.com/Automattic/woocommerce-payments/pull/3465#discussion_r763068901

In short, `create_intent` is still catching `API_Exception` but rather than returning a convoluted response array it's not returning a `WP_Error` object on failure and an intent id on success.

dpaun1985 reviewed (commented):
Overall looks good! I left some comments .

anu-rock committed:
Simplify code as per feedback

* Remove the 502 http code fallback
* Make capture method an optional param with a default value

anu-rock committed:
Add unit tests

anu-rock committed:
Fix the failing test

mordeth reviewed (approved):
Looks good to me!

dmvrtx commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,57 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+			'http_code' => $http_code,
+		];
+
+		if ( isset( $error_message ) ) {
```
It is better to throw/return an error early than making a response so complex. And a caller can use a simple `is_wp_error` call to verify if operation failed or not.

anu-rock replied:
Fixed in dca1ed4 ✅

dmvrtx reviewed (changes requested):
Good work, @anu-rock, and also good to split responsibilities between controller and gateway. There is an `$api_client` accessible in the controller directly, but I find your approach to be cleaner and it allows for simpler code re-use.

I have a few comments, but my main concern would be the gateway part. I believe that making it to return or throw an error will save some future developers from accidental mistakes 🙂

anu-rock committed:
Reduce the scope of try block only to error-prone code

anu-rock committed:
Refactor the payment method as a const for better readability

anu-rock committed:
Inline the single-use variables directly in method calls

anu-rock committed:
Simplify response and error handling in gateway

anu-rock committed:
Merge branch 'develop' into add/3431-create-intent-endpoint

anu-rock commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,48 @@ public function create_customer( $request ) {
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_terminal_intent( $request ) {
```
That's a good point. Fixed in 616fb1d ✅ 

> retrieve an order

`wc_get_order` uses `WC_Order_Factory` which internally handles any exceptions. As a result, it does not throw an exception to be caught here.

anu-rock committed:
Inline the amount variable in method call

Because:
(a) prepare_amount method is intuitively named and accepts only two arguments
(b) the amount variable is used only once

anu-rock commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,48 @@ public function create_customer( $request ) {
+		try {
+			$order_id        = $request['order_id'];
+			$payment_methods = [ 'card_present' ];
+			$capture_method  = 'manual';
```
Fixed via d3fb481 and db77cbb ✅

anu-rock commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,48 @@ public function create_customer( $request ) {
+	 */
+	public function create_terminal_intent( $request ) {
+		try {
+			$order_id        = $request['order_id'];
```
Fixed in dca1ed4 ✅

anu-rock commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +208,48 @@ public function create_customer( $request ) {
+
+			// Create a new intention.
+			$result = $this->gateway->create_intent( $order, $payment_methods, $capture_method );
+			if ( 'created' !== $result['status'] ) {
```
Fixed in dca1ed4 ✅

anu-rock commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,57 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+			$status    = 'created';
+			$http_code = 200;
+		} catch ( API_Exception $e ) {
+			$status        = 'failed';
```
Fixed in dca1ed4 ✅

anu-rock commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,57 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+		$response = [
+			'status'    => $status,
+			'id'        => ! empty( $intent ) ? $intent->get_id() : null,
+			'http_code' => $http_code,
```
Fixed in dca1ed4 ✅

anu-rock commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2443,6 +2443,57 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+	 *
+	 * @throws Exception - When an error occurs in intent creation.
+	 */
+	public function create_intent( WC_Order $order, array $payment_methods, string $capture_method = 'automatic' ) {
```
Fixed in dca1ed4 ✅

anu-rock commented:
@dpaun1985, @dmvrtx, @mordeth Thanks a tonne for your reviews. I've pushed new changes that hopefully address your feedback.

dmvrtx commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -199,4 +209,31 @@ public function create_customer( $request ) {
+		try {
+			// Create a new intention.
+			$result = $this->gateway->create_intent( $order, [ Payment_Method::CARD_PRESENT ], 'manual' );
+			if ( is_wp_error( $result ) ) {
```
This check is superfluous. [`rest_ensure_response`](https://developer.wordpress.org/reference/functions/rest_ensure_response/) will handle `WP_Error` for you.

kalessil replied:
+1, the whole try-body can be simplified to `return rest_ensure_response( $this->gateway->create_intent( $order, [ Payment_Method::CARD_PRESENT ], 'manual' ) );`

anu-rock replied:
Ya'll got it 205f7b4 😄 

P.S. I like to keep method call and return statements on separate lines 😉

kalessil replied:
Thanks!

dmvrtx reviewed (approved):
Thank you for your elaborate work on the review comments. There is one non-blocking comment, but otherwise good to go!

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2456,6 +2456,48 @@ public function schedule_order_tracking( $order_id, $order = null ) {
+	 *
+	 * @param WC_Order $order - Order based on which to create intent.
+	 * @param array    $payment_methods - A list of allowed payment methods. Eg. card, card_present.
+	 * @param ?string  $capture_method - Controls when the funds will be captured from the customer's account ("automatic" or "manual").
```
`@param ?string` perhaps?

anu-rock replied:
Isn't it the same already?

kalessil replied:
Nope, `?string` in PHP types is `string|null`. And we are requiring string only.

anu-rock replied:
Gotcha! Fixed here a313299 ✅

kalessil reviewed (approved):
LGTM! Also please check Valery's comment about possible simplification, I shared an example there.

anu-rock committed:
Remove the unnecessary wp error check

rest_ensure_response can handle WP_Error objects as well.

anu-rock committed:
Merge branch 'develop' into add/3431-create-intent-endpoint

anu-rock committed:
Fix capture_method type in doc comments
